### ISSUE TITLE
RubyGems `dpl` not working due to `gems` version

### ISSUE THREAD

**Issue by azeem-sajid-confiz** [travis-ci/travis-ci#8592]
Hi,

I'm trying to use Travis CI to release a gem using RubyGems provider but it is not working.

Though, using the following script with `provider: script` works fine:

```
#!/bin/sh

yes y | gem uninstall dpl
yes y | gem uninstall gems
gem install gems -v 0.8.3
gem install dpl -v 1.8.27
dpl --provider=rubygems --api-key=$API_KEY
```

The API key is not encrypted here. I tried with encrypted key but it does not work.

Can somebody guide me how to use encrypted key with this script?

I've used environment variable with Travis CI and it works fine.
But, the key is not encrypted in this case also.

Any relevant pointers would be helpful. Thanks!

**Comment by BanzaiMan** [comment]
Do you have a build log URL that shows the problem you are describing here?

**Comment by azeem-sajid-confiz** [comment]
@BanzaiMan 
Here's the build log: https://travis-ci.org/azeem-sajid-confiz/fog-azure-rm/builds/288894064

`.travis.yml` file: https://github.com/azeem-sajid-confiz/fog-azure-rm/blob/master/.travis.yml

I've gone through this thread also: [Unable to deploy new gem with latest gems](https://www.bountysource.com/issues/41312957-unable-to-deploy-new-gem-with-latest-gems), but, adding a `gem` and `gemspec` attributes under `deploy` didn't work either.

**Comment by BanzaiMan** [comment]
I don't think this is a problem with `dpl` or `gems`.

[The error](https://travis-ci.org/azeem-sajid-confiz/fog-azure-rm/builds/288894064#L4611-L4623) is:

```
/home/travis/.rvm/rubies/ruby-2.2.7/lib/ruby/2.2.0/json/common.rb:155:in `parse': 776: unexpected token at 'This rubygem could not be found.' (JSON::ParserError)
	from /home/travis/.rvm/rubies/ruby-2.2.7/lib/ruby/2.2.0/json/common.rb:155:in `parse'
	from /home/travis/.rvm/gems/ruby-2.2.7/gems/gems-1.0.0/lib/gems/client.rb:27:in `info'
	from /home/travis/.rvm/gems/ruby-2.2.7/gems/gems-1.0.0/lib/gems.rb:17:in `method_missing'
	from /home/travis/.rvm/gems/ruby-2.2.7/gems/dpl-1.8.43/lib/dpl/provider/rubygems.rb:28:in `check_app'
	from /home/travis/.rvm/gems/ruby-2.2.7/gems/dpl-1.8.43/lib/dpl/provider.rb:143:in `block in deploy'
	from /home/travis/.rvm/gems/ruby-2.2.7/gems/dpl-1.8.43/lib/dpl/cli.rb:41:in `fold'
	from /home/travis/.rvm/gems/ruby-2.2.7/gems/dpl-1.8.43/lib/dpl/provider.rb:141:in `deploy'
	from /home/travis/.rvm/gems/ruby-2.2.7/gems/dpl-1.8.43/lib/dpl/cli.rb:32:in `run'
	from /home/travis/.rvm/gems/ruby-2.2.7/gems/dpl-1.8.43/lib/dpl/cli.rb:7:in `run'
	from /home/travis/.rvm/gems/ruby-2.2.7/gems/dpl-1.8.43/bin/dpl:5:in `<top (required)>'
	from /home/travis/.rvm/gems/ruby-2.2.7/bin/dpl:23:in `load'
	from /home/travis/.rvm/gems/ruby-2.2.7/bin/dpl:23:in `<main>'
```

RubyGems is complaining about the information regarding `for-azure-rm-test` not found. https://rubygems.org/gems/for-azure-rm-test/versions

I believe that, currently, `dpl` is making the assumption that the gem already exists on rubygems.org before deploying with it.

Could you try pushing the first one manually?

**Comment by azeem-sajid-confiz** [comment]
Yes, I've already tried to push it manually and it worked! The gem was published.

**Comment by BanzaiMan** [comment]
Could you confirm that the deployment with `dpl` is now working?

**Comment by azeem-sajid-confiz** [comment]
Yes.
The manual deployment with `dpl` is successful.

I've used this command:
```
$ dpl --provider=rubygems --api-key=<api-key> --gemspec=<gemspec-file>
```
And, it works.

But, the deployment with Travis is not working again.
Here's the build log: https://travis-ci.org/azeem-sajid-confiz/fog-azure-rm/builds/289345746

**Comment by BanzaiMan** [comment]
Thanks for the information. I've reported the issue as https://github.com/rubygems/rubygems.org/issues/1683, as well as https://github.com/rubygems/gems/pull/41.

If neither of these issues are addressed soon, we will have to patch it up with `dpl`, so that the deployment can proceed. You can also try actually pushing the gem to rubygems.org.

Did your `dpl` command actually work? Why isn't there anything in https://rubygems.org/gems/for-azure-rm-test/versions?

**Comment by azeem-sajid-confiz** [comment]
Thank you!

Yes. The `dpl` command worked successfully when I tried it manually and it published this gem (`v0.0.3`): https://rubygems.org/gems/fog-azure-rm-test

For Travis deployment, with `v0.0.4`, it didn't work.
The build log I sent you earlier is for `v0.0.4` for the deployment that failed.

There was a typo in your link: 
https://rubygems.org/gems/for-azure-rm-test/versions
_______________________^

It is `fog`, not `for`:
https://rubygems.org/gems/fog-azure-rm-test/versions (This should work!)
_______________________^

**Comment by BanzaiMan** [comment]
Ah, thanks. Does `dpl` deployment work for you now?

**Comment by BanzaiMan** [comment]
By this, I mean, "on Travis CI", not with CLI.

**Comment by azeem-sajid-confiz** [comment]
No. It's not working on Travis CI.
I've restarted the build and it ended up with same errors.

Build log: https://travis-ci.org/azeem-sajid-confiz/fog-azure-rm/builds/289345746

**Comment by BanzaiMan** [comment]
I understand the confusion now. You have a typo here: https://github.com/azeem-sajid-confiz/fog-azure-rm/blob/4d9edfc0bbd5b12f91866d5fcdb2d81cfe3e2161/.travis.yml#L41

**Comment by azeem-sajid-confiz** [comment]
Oh, my bad. Thanks a lot!
I've corrected the typo and it passed the lookup phase.
But, it didn't publish the gem.

Here's the deployment log:

```
Deploying application
WARNING:  no email specified
WARNING:  pessimistic dependency on minitest (~> 5.8.4, development) may be overly strict
  if minitest is semantically versioned, use:
    add_development_dependency 'minitest', '~> 5.8', '>= 5.8.4'
WARNING:  open-ended dependency on simplecov (>= 0, development) is not recommended
  if simplecov is semantically versioned, use:
    add_development_dependency 'simplecov', '~> 0'
WARNING:  pessimistic dependency on codeclimate-test-reporter (~> 1.0.0, development) may be overly strict
  if codeclimate-test-reporter is semantically versioned, use:
    add_development_dependency 'codeclimate-test-reporter', '~> 1.0', '>= 1.0.0'
WARNING:  pessimistic dependency on fog-core (~> 1.43.0) may be overly strict
  if fog-core is semantically versioned, use:
    add_runtime_dependency 'fog-core', '~> 1.43', '>= 1.43.0'
WARNING:  prerelease dependency on azure-storage (= 0.11.5.preview) is not recommended
WARNING:  See http://guides.rubygems.org/specification-reference/ for help
  Successfully built RubyGem
  Name: fog-azure-rm-test
  Version: 0.0.4
  File: fog-azure-rm-test-0.0.4.gem
Access Denied. Please sign up for an account at https://rubygems.org
Already up-to-date!
Not currently on any branch.
Untracked files:
  (use "git add <file>..." to include in what will be committed)
	.bundle/
nothing added to commit but untracked files present (use "git add" to track)
Dropped refs/stash@{0} (272c43382f137194be3ced18bfcee2d8c54ef14d)
Done. Your build exited with 0.
```

I don't understand this part:
`Access Denied. Please sign up for an account at https://rubygems.org`

I've followed this tutorial to set up Travis CI for publishing gem:
https://docs.travis-ci.com/user/deployment/rubygems/

And, this for encrypting API key:
https://docs.travis-ci.com/user/encryption-keys/

The API key without encryption worked with RubyGems provider.
It also worked when it was stored as an environment variable (without encryption) for the script.

**Comment by BanzaiMan** [comment]
It sounds like your encryption went wrong. What command did you exactly use, in which directory? Often, you can remove ambiguity by passing `-r azeem-sajid-confiz/fog-azure-rm`:

    travis encrypt -r azeem-sajid-confiz/fog-azure-rm

**Comment by azeem-sajid-confiz** [comment]
In my repo directory, I've used this command:
```
$ travis encrypt API_key=<api-key> -r azeem-sajid-confiz/fog-azure-rm
```

And, then, put the encrypted key in `secure` under `api_key`:
```
deploy:
  provider: rubygems
  api_key:
    - secure: <encrypted-key>
  # ...
```

Let me try again with another repo.

**Comment by BanzaiMan** [comment]
The error comes from: https://github.com/travis-ci/dpl/blob/837da9ab01dc3a6981e9acb7a058091264db3f7f/lib/dpl/provider/rubygems.rb#L45:

```ruby
::Gems.push(File.new f)
```

which is a straightforward [POST](https://github.com/rubygems/gems/blob/v1.0.0/lib/gems/client.rb#L63) to the API endpoint. If that call is failing, I must think that your token is invalid.

That deployment failure does not fail the build is a known issue (for some deployment providers, of which RubyGems is one). Sorry.

**Comment by azeem-sajid-confiz** [comment]
Right. Got it. Thanks!

Finally, it worked when I encrypted the API key without a variable.
I'm closing this issue for being solved.
